### ISSUE TITLE
Intrinsic curiosity module (ICM) for exploration reward shaping

### ISSUE THREAD

**Issue by espenhk** [espenhk/gamer-ai#24]
## Summary

Add an **Intrinsic Curiosity Module (ICM)** as an optional reward shaping component in `rl/reward.py`. ICM generates an intrinsic reward signal proportional to the prediction error of a learned forward model, driving the agent to explore novel states rather than getting stuck in local loops.

## Motivation

Observations from current training runs:

1. **Early episode collapse**: policies frequently stall at the start line or drive in circles — the extrinsic `progress_weight` reward is sparse until the car actually moves forward
2. **Q-table sparsity**: `EpsilonGreedyPolicy` / `MCTSPolicy` visit a tiny fraction of states; tabular exploration is uniform and undirected
3. **Hill-climbing plateau**: `WeightedLinearPolicy` + `GeneticPolicy` spend many generations on near-zero-progress episodes

ICM provides a **dense, shaped intrinsic reward** by measuring how surprising each transition `(s, a, s')` is to a learned model. The agent is intrinsically motivated to visit states where its world model is worst — exactly the novel or challenging parts of the track.

### Architecture

ICM consists of two small MLPs (pure numpy, consistent with codebase):

**Inverse dynamics model** (IDM): predicts `a` from `(s, s')` — forces feature space to encode action-relevant information only:
```
ϕ(s), ϕ(s')  →  MLP  →  predicted action logits
```

**Forward model**: predicts next feature vector from current features + action:
```
[ϕ(s), a_onehot]  →  MLP  →  predicted ϕ(s')
```

**Feature encoder** ϕ: small linear/MLP projection of raw obs.

**Intrinsic reward**:
```
r_i = η * || ϕ(s') - ϕ̂(s') ||²
```
where `η` is a scale factor and `ϕ̂` is the forward model's prediction.

### Integration with `RewardCalculator`

Add to `RewardConfig` (`rl/reward.py`):

```yaml
# reward_config.yaml
icm_weight: 0.1          # scale of intrinsic reward (0 = disabled)
icm_hidden_size: 32
icm_feature_dim: 8       # compressed feature dimension
icm_lr: 0.001            # learning rate for ICM model updates
icm_beta: 0.2            # IDM vs forward model loss weighting
```

`RewardCalculator.compute()` gains an optional `curiosity_reward` component:

```python
if self.cfg.icm_weight > 0:
    r_intrinsic = self._icm.reward(prev_obs, action, curr_obs)
    self._icm.update(prev_obs, action, curr_obs)  # online model training
    total += self.cfg.icm_weight * r_intrinsic
```

### New class: `ICM` in `rl/reward.py` (or `rl/curiosity.py`)

```python
class ICM:
    def __init__(self, obs_dim, n_actions, feature_dim, hidden_size, lr, beta):
        ...

    def reward(self, s, a, s_next) -> float:
        """Return prediction error as intrinsic reward."""

    def update(self, s, a, s_next):
        """Online gradient update of encoder + IDM + forward model."""

    def reset_episode(self):
        """No-op for ICM (no per-episode state)."""
```

### Alternative: Random Network Distillation (RND)

A simpler alternative to ICM: a fixed random target network `f(s)` and a trained predictor network `g(s; θ)`. Intrinsic reward = `|| f(s') - g(s'; θ) ||²`. Requires no action conditioning and is cheaper to train. Could be offered as `curiosity_type: rnd` vs `curiosity_type: icm`.

## Files to change

| File | Change |
|------|--------|
| `rl/reward.py` | Add `ICM` class; extend `RewardConfig` + `RewardCalculator.compute()` |
| `rl/curiosity.py` | Optional: separate file for `ICM` + `RND` classes |
| `config/reward_config.yaml` | Document `icm_weight`, `icm_*` params (default 0 = off) |
| `tests/test_reward.py` | Unit test: ICM reward is higher for novel vs repeated transitions |

## Acceptance Criteria

- [ ] `icm_weight: 0` (default) produces identical reward to current `RewardCalculator` — fully backward compatible
- [ ] ICM updates its model weights online each step (no replay buffer required)
- [ ] Intrinsic reward is bounded and scaled appropriately relative to extrinsic rewards
- [ ] Works with all policy types (ICM is reward-side, policy-agnostic)
- [ ] Optional `curiosity_type: rnd` as a simpler alternative with same API
- [ ] Unit test: repeated identical transitions → decreasing intrinsic reward (model learns → less surprise)

## Expected Impact

- Faster initial exploration, especially for tabular policies (`EpsilonGreedy`, `MCTSPolicy`)
- Reduced dependence on carefully tuned `epsilon` / exploration constants
- May reduce cold-start episodes needed before hill-climbing finds a viable policy

## References

- Pathak et al. (2017). *Curiosity-driven Exploration by Self-supervised Prediction.* ICML 2017.
- Burda et al. (2018). *Exploration by Random Network Distillation.* arXiv:1810.12894

**Comment by debaa98** [comment]
is there any specific reward for contributers ?

**Comment by espenhk** [comment]
Hi @debaa98 ! This is really just a hobby project I made public to have something to reference in a portfolio. I'm happy to review a PR if you want to test it out and look at putting up some code, but I don't have any rewards to offer, no. It's really something I built (am still building, frankly) that I haven't put much work into preparing for open-sourcing.

Like I said, feel free to give it a shot though! Big NB: I just merged a massive refactor of the code, so the current issue desriptions as-is are probably off about where the code changes need to happen. I've refactored the codebase so extending the framework with multiple games should in theory be possible. That remains untested, though.
